iree/compiler/Conversion/Common/LinalgBufferizePass.cpp

```diff
@@ -540,21 +554,42 @@ LogicalResult preProcessInterfaceStoreTensorOp(
   return success();
 }
 
-/// Pre process linalg operations (on tensors) so that if the result of the
-/// linalg operation is mapped to a buffer, the out is mapped to the same buffer
-/// to make this an inplace update.
+/// Pre process linalg operations (on tensors) to propagate buffer assignment
+/// from results to operands wherever possible.
 LogicalResult preProcessLinalgOps(OpBuilder &b, linalg::LinalgOp op,
                                   BlockAndValueMapping &bvm) {
   if (!op.hasTensorSemantics()) return success();
+
   for (auto en :
        llvm::zip(op.getOperation()->getResults(), op.getOutputTensors())) {
     Value resultTensor = std::get<0>(en);
     Value outTensor = std::get<1>(en);
+    unsigned resultIndex = resultTensor.cast<OpResult>().getResultNumber();
     Value resultBuffer = bvm.lookupOrNull(resultTensor);
+
+    // If the result is mapped to a buffer, the corresponding output tensor can
+    // be mapped to the same buffer to make this an inplace update.
     if (resultBuffer && outTensor.hasOneUse()) {
       bvm.map(outTensor, resultBuffer);
     }
+
+    // If the output tensor is not actually used (for initialization) by this
+    // op, we can reuse the result tensor's buffer for some operand.
+    if (!op.payloadUsesValueFromOutputOperandIndex(resultIndex)) {
```

Comment:
Can you give me an example of why this is needed. This is not the intended contract of `out` tensor. Think of TiedOpInterface. The `out` operand is tied to the result. Tie-ing it to one of the operands is strange.
Some use case here would help, cause I cant think of a reason to have this if you are only fusing named ops with consumers.

Reply:
It's for eliding intermediate buffers needed between `linalg.conv` and its consumer `linalg.generic`. Without this, for the following fused IR in the same dispatch region:

```mlir
scf.for
  scf.for
    scf.for
      %init0 = linalg.init_tensor
      %init1 = linalg.init_tensor
      %fill = linalg.fill(%init1, ...)
      %conv = linalg.conv_2d_input_nhwc_filter_hwcf outs(%fill)
      %generic = linalg.generic ins(%conv, ...) outs(%init0)
```

After bufferization, we'll see an allocation for the result for `%conv`:


```mlir
scf.for
  scf.for
    scf.for
      %subview = memref.subview
      %alloc = memref.alloc
      linalg.fill(%alloc, ...)
      linalg.conv_2d_input_nhwc_filter_hwcf outs(%alloc)
      %generic = linalg.generic ins(%alloc, ...) outs(%subview)
```

The allocation is breaking the lowering to GPU and it's also unnecessary as we can just directly use `%subview`. The connection here is to basically tell the producer `%conv` to reuse the buffer for `%subview`.

I'm also feeling a bit uneasy about this.. But this seems the best way to make the connection, without performing more complicated analysis afterwards to elide the allocation. If you think there are better ways to handle this, I'm certainly happy to try out.

Reply:
Yeah, I suspected this was the case. This can get tricky though. Best approach to try here is to mark these as TiedOperands at dispatch region creation (and use that information here).
The advantage there would be if we can tie the operands, we just bail on the fusion. So we wont hit a case where at a later stage we have to allocate a buffer.

Reply:
I'm not sure how we can mark them as tied in dispatch region formation, given they are *internal* tensors to the `flow.dispatch.workgroups` op. The result of `%conv` is only used by the consumer `%generic`. It does not escape to the `flow.dispatch.workgroups` interface so there is no block argument for it.